signer/src/transaction_coordinator.rs

```diff
@@ -732,9 +737,102 @@ where
             .increment(1);
         }
 
+        for withdrawal in rejected_withdrawals {
+            let withdrawal_id = withdrawal.qualified_id();
+            let fut = self.construct_and_sign_withdrawal_reject(
+                chain_tip,
+                &wallet,
+                bitcoin_aggregate_key,
+                withdrawal,
+            );
+            if let Err(error) = fut.await {
+                tracing::warn!(
+                    %error,
+                    ?withdrawal_id,
+                    "could not construct and sign withdrawal reject"
+                );
+            }
+        }
+
         Ok(())
     }
 
+    async fn construct_and_sign_withdrawal_reject(
+        &mut self,
+        chain_tip: &model::BitcoinBlockRef,
+        wallet: &SignerWallet,
+        bitcoin_aggregate_key: &PublicKey,
+        request: model::WithdrawalRequest,
+    ) -> Result<(), Error> {
+        let db = self.context.get_storage();
+        let stacks = self.context.get_stacks_client();
+        let bitcoin = self.context.get_bitcoin_client();
+        let deployer = self.context.config().signer.deployer;
+
+        let is_completed = stacks
+            .is_withdrawal_completed(&deployer, request.request_id)
+            .await?;
+
+        if is_completed {
+            // The request is already completed according to the contract
+            return Ok(());
+        }
+
+        let valid_outputs = db
+            .get_valid_withdrawal_outputs(&request.qualified_id())
+            .await?;
+
+        for output in valid_outputs {
+            let mempool_entry = bitcoin
+                .get_mempool_entry(&output.bitcoin_txid.into())
+                .await?;
+            if mempool_entry.is_some() {
+                // There's an output for this request in the mempool, let's wait
+                return Ok(());
+            }
+        }
```

Comment:
Should we also check if the tx is confirmed? eg in case it's not in the mempool because it was just accepted but we are still missing the notification from zmq.

Reply:
Yeah, couldn't hurt right. And it's an edge case that we need to consider. I think `BitcoinInteract::get_tx` checks for the mempool or a block.

Reply:
Done in [bc3ea8f](https://github.com/stacks-network/sbtc/pull/1398/commits/bc3ea8f216f14087f02b6c731ad7348e57e9d1bf).